File: ios/objNative/JSThemeUniqueItem.m
//
//  JSThemeUniqueItem.m
//  Supermap
//
//  Created by supermap on 2018/8/9.
//  Copyright © 2018年 Facebook. All rights reserved.
//

#import "JSThemeUniqueItem.h"
#import "JSObjManager.h"
#import "SuperMap/ThemeUniqueItem.h"
@implementation JSThemeUniqueItem
RCT_EXPORT_MODULE();
RCT_REMAP_METHOD(createObj, createObjResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        ThemeUniqueItem* item = [[ThemeUniqueItem alloc]init];
        NSInteger itemId = (NSInteger)item;
        [JSObjManager addObj:item];
        resolve(@(itemId).stringValue);
    } @catch (NSException *exception) {
        reject(@"JSThemeUniqueItem",@"createObj expection",nil);
    }
}
/**
 * 返回单值专题图子项的名称
 * @param themeUniqueItemId
 */
RCT_REMAP_METHOD(getCaption,getCaptionById:(NSString*)themeUniqueItemId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        ThemeUniqueItem* themeUniqueItem = [JSObjManager getObjWithKey:themeUniqueItemId];
        NSString* caption = themeUniqueItem.mCaption;
        resolve(caption);
    } @catch (NSException *exception) {
        reject(@"JSThemeUniqueItem",@"getCaption expection",nil);
    }
}

/**
 * 返回单值专题图子项所对应的显示风格
 * @param themeUniqueItemId
 */
RCT_REMAP_METHOD(getStyle,getStyleById:(NSString*)themeUniqueItemId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        ThemeUniqueItem* themeUniqueItem = [JSObjManager getObjWithKey:themeUniqueItemId];
        GeoStyle* style = themeUniqueItem.mStyle;
        NSInteger styleId = (NSInteger)style;
        [JSObjManager addObj:style];
        resolve(@(styleId).stringValue);
    } @catch (NSException *exception) {
        reject(@"JSThemeUniqueItem",@"getStyle expection",nil);
    }
}
/**
 * 返回单值专题图子项的单值
 * @param themeUniqueItemId
 */
RCT_REMAP_METHOD(getUnique,getUniqueById:(NSString*)themeUniqueItemId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        ThemeUniqueItem* themeUniqueItem = [JSObjManager getObjWithKey:themeUniqueItemId];
        NSString* value = themeUniqueItem.mUnique;
        resolve(value);
    } @catch (NSException *exception) {
        reject(@"JSThemeUniqueItem",@"getUnique expection",nil);
    }
}

/**
 * 返回单值专题图子项是否可见
 * @param themeUniqueItemId
 */

RCT_REMAP_METHOD(isVisible,isVisibleById:(NSString*)themeUniqueItemId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        ThemeUniqueItem* themeUniqueItem = [JSObjManager getObjWithKey:themeUniqueItemId];
        BOOL visible = themeUniqueItem.mIsVisible;
        NSNumber* number = [NSNumber numberWithBool:visible];
        resolve(number);
    } @catch (NSException *exception) {
        reject(@"JSThemeUniqueItem",@"isVisible expection",nil);
    }
}
/**
 * 设置单值专题图子项的名称
 * @param themeUniqueItemId
 * @param value
 */
RCT_REMAP_METHOD(setCaption,setCaptionById:(NSString*)themeUniqueItemId value:(NSString*)value resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        ThemeUniqueItem* themeUniqueItem = [JSObjManager getObjWithKey:themeUniqueItemId];
        themeUniqueItem.mCaption = value;
        NSNumber* number = [NSNumber numberWithBool:YES];
        resolve(number);
    } @catch (NSException *exception) {
        reject(@"JSThemeUniqueItem",@"setCaption expection",nil);
    }
}

/**
 * 设置单值专题图子项所对应的显示风格
 * @param themeUniqueItemId
 * @param styleId
 */
RCT_REMAP_METHOD(setStyle,setStyleById:(NSString*)themeUniqueItemId styleId:(NSString*)styleId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        ThemeUniqueItem* themeUniqueItem = [JSObjManager getObjWithKey:themeUniqueItemId];
        GeoStyle *style = [JSObjManager getObjWithKey:styleId];
        themeUniqueItem.mStyle = style;
        NSNumber* number = [NSNumber numberWithBool:YES];
        resolve(number);
    } @catch (NSException *exception) {
        reject(@"JSThemeUniqueItem",@"setStyle expection",nil);
    }
}
/**
 * 设置单值专题图子项的单值
 * @param themeUniqueItemId
 * @param value
 */
RCT_REMAP_METHOD(setUnique,setUniqueById:(NSString*)themeUniqueItemId value:(NSString*)value resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        ThemeUniqueItem* themeUniqueItem = [JSObjManager getObjWithKey:themeUniqueItemId];
        themeUniqueItem.mUnique = value;
        NSNumber* number = [NSNumber numberWithBool:YES];
        resolve(number);
    } @catch (NSException *exception) {
        reject(@"JSThemeUniqueItem",@"setUnique expection",nil);
    }
}

/**
 * 设置单值专题图子项是否可见
 * @param themeUniqueItemId
 * @param value
 */
RCT_REMAP_METHOD(setVisible,setVisibleById:(NSString*)themeUniqueItemId value:(BOOL)value resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        ThemeUniqueItem* themeUniqueItem = [JSObjManager getObjWithKey:themeUniqueItemId];
        themeUniqueItem.mIsVisible = value;
        NSNumber* number = [NSNumber numberWithBool:YES];
        resolve(number);
    } @catch (NSException *exception) {
        reject(@"JSThemeUniqueItem",@"setVisible expection",nil);
    }
}
/**
 * 返回单值专题图子项格式化字符串
 * @param themeUniqueItemId
 */
RCT_REMAP_METHOD(toString,toStringById:(NSString*)themeUniqueItemId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        ThemeUniqueItem* themeUniqueItem = [JSObjManager getObjWithKey:themeUniqueItemId];
        NSString* string = themeUniqueItem.toString;
        resolve(string);
    } @catch (NSException *exception) {
        reject(@"JSThemeUniqueItem",@"toString expection",nil);
    }
}
@end

File: android/src/main/java/com/supermap/rnsupermap/JSThemeUniqueItem.java
package com.supermap.rnsupermap;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.supermap.data.GeoStyle;
import com.supermap.mapping.ThemeUniqueItem;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

public class JSThemeUniqueItem extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = "JSThemeUniqueItem";
    protected static Map<String, ThemeUniqueItem> m_ThemeUniqueItemList = new HashMap();
    ThemeUniqueItem m_ThemeUniqueItem;

    public JSThemeUniqueItem(ReactApplicationContext context) {
        super(context);
    }

    public static ThemeUniqueItem getObjFromList(String id) {
        return m_ThemeUniqueItemList.get(id);
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    public static String registerId(ThemeUniqueItem obj) {
        for (Map.Entry entry : m_ThemeUniqueItemList.entrySet()) {
            if (obj.equals(entry.getValue())) {
                return (String) entry.getKey();
            }
        }

        Calendar calendar = Calendar.getInstance();
        String id = Long.toString(calendar.getTimeInMillis());
        m_ThemeUniqueItemList.put(id, obj);
        return id;
    }

    @ReactMethod
    public void createObj(Promise promise){
        try{
            ThemeUniqueItem item = new ThemeUniqueItem();
            String id = registerId(item);

            promise.resolve(id);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    /**
     * 返回单值专题图子项的名称
     * @param themeUniqueItemId
     * @param promise
     */
    @ReactMethod
    public void getCaption(String themeUniqueItemId, Promise promise){
        try{
            ThemeUniqueItem themeUniqueItem = getObjFromList(themeUniqueItemId);
            String caption = themeUniqueItem.getCaption();
            promise.resolve(caption);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    /**
     * 返回单值专题图子项所对应的显示风格
     * @param themeUniqueItemId
     * @param promise
     */
    @ReactMethod
    public void getStyle(String themeUniqueItemId, Promise promise){
        try{
            ThemeUniqueItem themeUniqueItem = getObjFromList(themeUniqueItemId);
            GeoStyle style = themeUniqueItem.getStyle();
            String styleId = JSGeoStyle.registerId(style);

            promise.resolve(styleId);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    /**
     * 返回单值专题图子项的单值
     * @param themeUniqueItemId
     * @param promise
     */
    @ReactMethod
    public void getUnique(String themeUniqueItemId, Promise promise){
        try{
            ThemeUniqueItem themeUniqueItem = getObjFromList(themeUniqueItemId);
            String value = themeUniqueItem.getUnique();

            promise.resolve(value);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    /**
     * 返回单值专题图子项是否可见
     * @param themeUniqueItemId
     * @param promise
     */
    @ReactMethod
    public void isVisible(String themeUniqueItemId, Promise promise){
        try{
            ThemeUniqueItem themeUniqueItem = getObjFromList(themeUniqueItemId);
            boolean visible = themeUniqueItem.isVisible();
            promise.resolve(visible);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    /**
     * 设置单值专题图子项的名称
     * @param themeUniqueItemId
     * @param value
     * @param promise
     */
    @ReactMethod
    public void setCaption(String themeUniqueItemId, String value, Promise promise){
        try{
            ThemeUniqueItem themeUniqueItem = getObjFromList(themeUniqueItemId);
            themeUniqueItem.setCaption(value);
            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    /**
     * 设置单值专题图子项所对应的显示风格
     * @param themeUniqueItemId
     * @param styleId
     * @param promise
     */
    @ReactMethod
    public void setStyle(String themeUniqueItemId, String styleId, Promise promise){
        try{
            ThemeUniqueItem themeUniqueItem = getObjFromList(themeUniqueItemId);
            GeoStyle style = JSGeoStyle.getObjFromList(styleId);
            themeUniqueItem.setStyle(style);
            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    /**
     * 设置单值专题图子项的单值
     * @param themeUniqueItemId
     * @param value
     * @param promise
     */
    @ReactMethod
    public void setUnique(String themeUniqueItemId, String value, Promise promise){
        try{
            ThemeUniqueItem themeUniqueItem = getObjFromList(themeUniqueItemId);
            themeUniqueItem.setUnique(value);
            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    /**
     * 设置单值专题图子项是否可见
     * @param themeUniqueItemId
     * @param value
     * @param promise
     */
    @ReactMethod
    public void setVisible(String themeUniqueItemId, boolean value, Promise promise){
        try{
            ThemeUniqueItem themeUniqueItem = getObjFromList(themeUniqueItemId);
            themeUniqueItem.setVisible(value);
            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    /**
     * 返回单值专题图子项格式化字符串
     * @param themeUniqueItemId
     * @param promise
     */
    @ReactMethod
    public void toString(String themeUniqueItemId, Promise promise){
        try{
            ThemeUniqueItem themeUniqueItem = getObjFromList(themeUniqueItemId);
            String string = themeUniqueItem.toString();
            promise.resolve(string);
        }catch (Exception e){
            promise.reject(e);
        }
    }
}

